tokio/src/task/scope.rs

```diff
@@ -0,0 +1,439 @@
+    pub fn detached() -> Self {
+        Self::with_config(ScopeConfigBuilder::detached().build().expect("Default config can not fail"))
+    }
+
+    /// Creates a `Scope` with the given configuration
+    pub fn with_config(config: ScopeConfig) -> Self {
+        Self {
+            config
+        }
+    }
+
+    /// Creates a [`scope`] with custom options
+    ///
+    /// The method behaves like [`scope`], but the cancellation and `Drop` behavior
+    /// for the [`scope`] are configurable. See [`ScopeConfig`] for details.
+    pub async fn enter<F, Fut, R>(self, scope_func: F) -> R
+    where
+        F: FnOnce(ScopeHandle) -> Fut,
+        Fut: Future<Output = R> + Send,
+    {
+        let scope_state = ScopeState::new(self.config);
+        let wait_fut = scope_state.wait_group.wait_future();
+
+        // This guard will be called be executed if the scope gets dropped while
+        // it is still executing.
+        let mut wait_for_tasks_guard = WaitForTasksToJoinGuard {
+            wait_group: &scope_state.wait_group,
+            enabled: true,
+            drop_behavior: scope_state.config.drop_behavior,
+        };
+
+        let scoped_result = {
+            // This guard will call `.cancel()` on the `CancellationToken` we
+            // just created.
+            let _cancel_guard = CancelTasksGuard {
+                scope: &scope_state.config.cancellation_token,
+            };
+
+            let handle = ScopeHandle {
+                scope: scope_state.clone(),
+            };
+
+            // Execute the scope handler, which gets passed a handle to the newly
+            // created scope
+            scope_func(handle).await
+        };
+
+        // Wait for all remaining tasks inside the scope to complete
+        wait_fut.await;
+
+        // The tasks have completed. We do not need to wait for them to complete
+        // in the `Drop` guard.
+        wait_for_tasks_guard.disarm();
+
+        scoped_result
+    }
+}
+
+/// Creates a task scope with default options.
+///
+/// The `scope` allows to spawn child tasks so that the lifetime of child tasks
+/// is constrained within the scope.
+///
+/// A closure which accepts a [`ScopeHandle`] object and which returns a [`Future`]
+/// needs to be passed to `scope`. The [`ScopeHandle`] can be used to spawn child
+/// tasks.
+///
+/// If the provided `Future` had been polled to completion, all child tasks which
+/// have been spawned via the `ScopeHandle` will be cancelled.
+///
+/// `scope` returns a [`Future`] which should be awaited. The `await` will only
+/// complete once all child tasks that have been spawned via the provided
+/// [`ScopeHandle`] have joined. Thereby the `scope` does not allow child tasks
+/// to outlive their parent task, as long as the `scope` is awaited.
+pub async fn scope<F, Fut, R>(scope_func: F) -> R
+where
+    F: FnOnce(ScopeHandle) -> Fut,
+    Fut: Future<Output = R> + Send,
+{
+    Scope::detached().enter(scope_func).await
```

Comment:
In your other PR about "implicit scopes", this would attach to the parent, but in this PR, this is supposed to be a global/detached scope, yes?

Reply:
The "implicit scopes" change would also get a method where you can spawn a task which is not directly attached to the parent to opt-out of structured concurrency if you want to. Currently both options would then spawn a "global" task, but I think in the end both variants should then spawn the task into the "runtime scope", which is the top level scope. Thereby shutting down the runtime would automatically cancel the remaining tasks and wait for them to complete. This however requires a change to the runtime structs to include this top level scope, which is not included yet.